File: 1406-stone-game-iii/1406-stone-game-iii.java
class Solution {
    int solve(int index, int[] stones, int[] dp){
        
        if(index >= stones.length)
            return 0;
        
        if(dp[index] != -1)
            return dp[index];
        
        int result = Integer.MIN_VALUE;
        int sum = 0;
        
        for(int i=index; i<index+3 && i<stones.length; i++){
            
            sum += stones[i];
            result = Math.max(result, sum - solve(i+1, stones, dp));
        }
        
        return dp[index] = result;
    }
    
    int tabulation(int[] stones){
        
        int n = stones.length;
        
        int[] dp = new int[n+1];
        
        for(int index=n-1; index>=0; index--){
            
            int sum = 0;

            for(int i=index; i<index+3 && i<stones.length; i++){

                sum += stones[i];
                dp[i] = Math.max(dp[i], sum - dp[i+1]);
            }
        }
        
        return dp[0];
    }
    
    public String stoneGameIII(int[] stones) {
        
        int n = stones.length;
        
        int[] dp = new int[n];
        Arrays.fill(dp, -1);
        
        int diff = solve(0, stones, dp);
        
        if(diff > 0){
            return "Alice";
        }
        else if(diff < 0){
            return "Bob";
        }
        
        return "Tie";
    }
}

File: 1406-stone-game-iii/1406-stone-game-iii.cpp
class Solution {
public:
    int solve(int index, bool alice_turn, vector<int>& stoneValue,vector<vector<int>>& dp){
        
        if(index >= stoneValue.size())
            return 0;
        
        if(dp[index][alice_turn] != -1)
            return dp[index][alice_turn];
        
        int ans;
        int sum=0;
        
        if(alice_turn == true){
            
            ans = -1e9;
            for(int i=index; i<index + 3 && i<stoneValue.size(); i++){
                
                sum += stoneValue[i];
                ans = max(ans, sum + solve(i+1, false, stoneValue, dp));
            }
            
            return dp[index][alice_turn] = ans;
        }
        else{
            
            ans = 1e9;
            for(int i=index; i<index + 3 && i<stoneValue.size(); i++){
                
                ans = min(ans, solve(i+1, true, stoneValue, dp));
            }
            
            return dp[index][alice_turn] = ans;
        }
    }
    
    string stoneGameIII(vector<int>& stoneValue) {
        
        int n = stoneValue.size();
        
        vector<vector<int>> dp(n, vector<int> (2, -1));
        int alice_points = solve(0, true, stoneValue, dp);
        
        
        int bob_points = accumulate(begin(stoneValue), end(stoneValue), 0) - alice_points;
        
        if(alice_points > bob_points){
            return "Alice";
        }
        else if(alice_points < bob_points){
            return "Bob";
        }
        else{
            return "Tie";
        }
    }
};